src/sinks/datadog_archives.rs

```diff
@@ -0,0 +1,571 @@
+        static ref KEY_PREFIX: Template =
+            Template::try_from("/dt=%Y%m%d/hour=%H/").expect("invalid object key format");
+    }
+    let key = KEY_PREFIX
+        .render_string(&event)
+        .map_err(|error| {
+            emit!(TemplateRenderingFailed {
+                error,
+                field: Some("key_prefix"),
+                drop_event: true,
+            });
+        })
+        .ok()?;
+
+    let mut log = event.into_log();
+
+    log.insert("_id", generate_log_id());
+
+    let timestamp = log
+        .remove(crate::config::log_schema().timestamp_key())
+        .unwrap_or_else(|| chrono::Utc::now().timestamp_millis().into());
+    log.insert(
+        "date",
+        timestamp
+            .as_timestamp()
+            .cloned()
+            .unwrap_or_else(chrono::Utc::now)
+            .to_rfc3339_opts(SecondsFormat::Millis, true),
+    );
+    if let Some(message) = log.remove(crate::config::log_schema().message_key()) {
+        log.insert("message", message);
+    }
+    if let Some(message) = log.remove(crate::config::log_schema().host_key()) {
+        log.insert("host", message);
+    }
+
+    let mut attributes = BTreeMap::new();
+    let custom_attributes: Vec<String> = log
+        .keys()
+        .filter(|path| !RESERVED_ATTRIBUTES.contains(path.as_str()))
+        .collect();
+    for path in custom_attributes {
+        if let Some(value) = log.remove(&path) {
+            attributes.insert(path, value);
+        }
+    }
+    log.insert("attributes", attributes);
+
+    let bytes = serde_json::to_vec(&log)
+        .map(|mut b| {
+            b.push(b'\n');
+            b
+        })
+        .expect("Failed to encode event as json, this is a bug!");
+
+    Some(EncodedEvent {
+        item: PartitionInnerBuffer::new(bytes, key.into()),
+        finalizers: log.metadata_mut().take_finalizers(),
+    })
+}
+
+/// Generates a unique ID compatible with DD
+fn generate_log_id() -> String {
+    let mut id = BytesMut::with_capacity(18);
+    // timestamp in millis - 6 bytes
+    let now = Utc::now();
+    id.put_int(now.timestamp_millis(), 6);
+    // namespace + version, both 0 - 2 bytes
+    id.put_i16(0);
+
+    // host-based number - 5 bytes
+    lazy_static! {
+        static ref HOST_NUMBER: Vec<u8> = host_number();
+    }
+    id.put_slice(&HOST_NUMBER[..]);
+
+    // counter - 5 bytes
+    id.put_u8(0);
+    // one padding byte
+    // 4 bytes for the counter should be more than enough - it should be unique for 1 millisecond only
+    lazy_static! {
+        static ref COUNTER: CounterU32 = CounterU32::new(0);
+    }
+    id.put_u32(COUNTER.inc()); // 4 bytes
+
+    base64::encode(id.freeze())
+}
+
+/// Returns 5 first bytes of the MAC address XOR-ed with random bytes
+fn host_number() -> Vec<u8> {
+    extern crate mac_address;
+    use mac_address::get_mac_address;
+    use rand::Rng;
+
+    let mac = get_mac_address().unwrap().unwrap().bytes();
+    let mut rng = rand::thread_rng();
```

Comment:
This queries the thread local state for an RNG. Similar to my comments above please refactor this function into a structure and store the RNG you need as a part of that struct state.

Reply:
Refactored. Now the host number is stored in a struct.